File: built-in plugins dsp/mb-brass-trumpet.cpp
/**
 * MB Trumpet
 * Category : instrument
 * Type     : brass
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Bright solo trumpet with expressive dynamics
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_BRASS_TRUMPET_H
#define MB_BRASS_TRUMPET_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbBrassTrumpet : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-brass-trumpet";
    static constexpr const char* PLUGIN_NAME    = "MB Trumpet";
    static constexpr const char* PLUGIN_TYPE    = "brass";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float brightness = 0.7f;  // range [0, 1]
    float vibrato = 0.3f;  // range [0, 1]
    float mute = 0f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbBrassTrumpet() = default;
    ~MbBrassTrumpet() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.brightness = std::clamp(params.brightness, 0f, 1f);
        params.vibrato = std::clamp(params.vibrato, 0f, 1f);
        params.mute = std::clamp(params.mute, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Trumpet
        return input;
    }
};

#endif // MB_BRASS_TRUMPET_H

File: built-in plugins dsp/mb-brass-trumpet.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbBrassTrumpetPlugin: PluginDefinition = {
  id: "mb-brass-trumpet",
  slug: "mb-brass-trumpet",
  name: "MB Trumpet",
  category: "instrument",
  type: "brass" as any,
  version: "1.0.0",
  description: "Bright solo trumpet with expressive dynamics",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sawtooth", detune: 0, gain: 0.6 },
    { type: "square", detune: 0, gain: 0.3 },
    { type: "sine", detune: 1200, gain: 0.1 },
  ],
  envelope: { attack: 0.03, decay: 0.2, sustain: 0.85, release: 0.15 },
  parameters: [
    {
      id: "brightness",
      name: "Brightness",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "vibrato",
      name: "Vibrato",
      type: "float",
      defaultValue: 0.3,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "mute",
      name: "Mute",
      type: "float",
      defaultValue: 0,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { brightness: 0.7, vibrato: 0.3, mute: 0, volume: 0.8 },
};

export default MbBrassTrumpetPlugin;
